testing/multi-env-test-engine/src/main/java/org/projectnessie/junit/engine/MultiEnvTestFilter.java

```diff
@@ -47,11 +47,8 @@ private Optional<Class<?>> classFor(TestDescriptor object) {
   }
 
   private FilterResult filter(Class<?> testClass, UniqueId id) {
-    // Use the static extension data collected during the discovery phase.
-    // It is possible to reload extensions based of class objects from test descriptors,
-    // however that would add unnecessary overhead.
-    MultiEnvExtensionRegistry registry = registry();
-
+    // Keep separate registry from MultiEnvTestEngine in case we are running on another engine and still need to filter.
+    registry.registerExtensions(testClass);
     if (id.getEngineId().map("junit-jupiter"::equals).orElse(false)) {
       if (registry.stream(testClass).findAny().isPresent()) {
```

Comment:
WDYT about moving `segmentType()` to an annotation on `MultiEnvTestExtension`? Then a registry instance will not be required here at all. The `stream` method could probably be refactored to offer a `static` part for discovering annotations.

Reply:
Interesting, I'll give this a shot